pkg/proxy/proxy/transparent.go

```diff
@@ -115,31 +120,82 @@ func NewTransparentProxyService(p *goproxy.ProxyHttpServer, ca *tls.Certificate,
 	}
 }
 
+func (t TransparentProxyService) Shutdown(ctx context.Context) error {
+	if t.adminShutdown != nil {
+		if err := t.adminShutdown(ctx); err != nil {
+			return err
+		}
+	}
+	if t.httpShutdown != nil {
+		if err := t.httpShutdown(ctx); err != nil {
+			return err
+		}
+	}
+	if t.tlsShutdown != nil {
+		if err := t.tlsShutdown(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // ProxyHTTP serves an endpoint that transparently redirects HTTP connections to the proxy server.
 // This endpoint also explicitly (i.e. non-transparently) proxies HTTP and TLS connections.
-func (t TransparentProxyService) ProxyHTTP(addr string) {
+func (t *TransparentProxyService) ProxyHTTP(addr string) {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		log.Fatalf("Error listening for http connections - %v", err)
 	}
-	if err := http.Serve(ln, t.Proxy); err != nil {
-		log.Fatalln(err)
+	server := &http.Server{
+		Addr:    addr,
+		Handler: t.Proxy,
+	}
+	t.httpShutdown = func(ctx context.Context) error { return server.Shutdown(ctx) }
+	if err := server.Serve(ln); err != nil {
+		log.Println(err)
 	}
 }
 
 // ProxyTLS serves an endpoint that transparently redirects TLS connections to the proxy server.
-func (t TransparentProxyService) ProxyTLS(addr string) {
+func (t *TransparentProxyService) ProxyTLS(addr string) {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		log.Fatalf("Error listening for https connections - %v", err)
 	}
+	var inflightRequests sync.WaitGroup
+	t.tlsShutdown = func(ctx context.Context) error {
+		ch := make(chan struct{})
+		errChan := make(chan error)
+		go func() {
+			if err := ln.Close(); err != nil {
+				errChan <- err
+				return
+			}
+			log.Println("Waiting for in-flight requests to complete...")
+			inflightRequests.Wait()
+			close(ch)
```

Comment:
I think you can just send nil to the errChan here instead and get rid of ch. as a result you could rename if you wanted.